kv/storage: plumb contexts to DistSender, Node, Store, TxnCoordSender

RaduBerinde opened the pull request:
Tracers are removed, we now use the context to get the Tracer.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/8625)

<!-- Reviewable:end -->

andreimatei commented:
One thing that bothers me, as you know, is the multiple calls to `NewTracer`, even if I think most of them are in tests. Let's discuss it with @tschottdorf when he's back.

And please see the comments in `txn_coord_sender` about whether that guy should have a base context at all.

Otherwise <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> , God's work.

---

Review status: 0 of 11 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[kv/txn_coord_sender.go, line 148 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8625#-KPTJmcgzRsCQhz_U-oH:-KPTJmcgzRsCQhz_U-oI:b-5gv42t) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca6615357cff9ed65c813b50d79c60d75b4023f8/kv/txn_coord_sender.go#L148)):*
> ```Go
> // clears accumulated write intents for the transaction.
> type TxnCoordSender struct {
> 	ctx               context.Context
> ```

why does `TxnCoordSender` need a "base" context? As opposed to taking the ctx as an arg for all operations? In particular, you add a TODO in `Send()` to use the ctx that's passed in. Why can't we start doing that in this PR?

---

*[kv/txn_coord_sender.go, line 174 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8625#-KPTJ_KjE3YzNrvDmV-l:-KPTJ_KkcUccXYb2sgoj:bcrlz0v) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca6615357cff9ed65c813b50d79c60d75b4023f8/kv/txn_coord_sender.go#L174)):*
> ```Go
> 	txnMetrics TxnMetrics,
> ) *TxnCoordSender {
> 	if ctx == nil || tracing.TracerFromCtx(ctx) == nil {
> ```

check the ctx to not be cancellable too?

---

*[kv/txn_coord_sender.go, line 272 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8625#-KPTKCEpnPo2LxFy7yZP:-KPTKCEpnPo2LxFy7yZQ:bmc3ijw) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca6615357cff9ed65c813b50d79c60d75b4023f8/kv/txn_coord_sender.go#L272)):*
> ```Go
> 		// TODO(radu): once contexts are plumbed correctly, we should use the Tracer
> 		// from ctx.
> 		tracer := tracing.TracerFromCtx(tc.ctx)
> ```

so you use the span from one context and the tracer from another. Is that cool?

---

*[kv/txn_coord_sender.go, line 590 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8625#-KPTLLhrvpniwcQSZYpr:-KPTLLhrvpniwcQSZYps:baole7o) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca6615357cff9ed65c813b50d79c60d75b4023f8/kv/txn_coord_sender.go#L590)):*
> ```Go
> 	// TODO(tschottdorf): this should join to the trace of the request
> 	// which starts this goroutine.
> 	sp := tracing.TracerFromCtx(tc.ctx).StartSpan(opHeartbeatLoop)
> ```

shouldn't this use the `ctx` method arg?

---

*[kv/txn_coord_sender.go, line 667 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8625#-KPTLDRQ2sDr7j7GPyhn:-KPTLDRR-PkaARFny4nn:bdj8qlt) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca6615357cff9ed65c813b50d79c60d75b4023f8/kv/txn_coord_sender.go#L667)):*
> ```Go
> 		if _, pErr := tc.wrapped.Send(context.Background(), ba); pErr != nil {
> 			if log.V(1) {
> 				log.Warningf(tc.ctx, "abort due to inactivity failed for %s: %s ", txn, pErr)
> ```

shouldn't `tryAsyncAbort` be passed in a context related to the txn it's working on?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8625#-:-KPTLVsa5Vaokx-L_xGx:b-po3lls)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
TFTR! Note that I'm not adding any calls, just moving them around. I agree they should go away and I plan to work on that, but I'm taking small steps.

In a "real" server, everyone should be using the Tracer set up by server.go.

---

Review status: 0 of 11 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[kv/txn_coord_sender.go, line 148 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8625#-KPTJmcgzRsCQhz_U-oH:-KPTN2e-2K_28hsAsIDO:b-iu3a14) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca6615357cff9ed65c813b50d79c60d75b4023f8/kv/txn_coord_sender.go#L148)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> why does `TxnCoordSender` need a "base" context? As opposed to taking the ctx as an arg for all operations? In particular, you add a TODO in `Send()` to use the ctx that's passed in. Why can't we start doing that in this PR?
</details>
Well there are still logs that are not associated to any operations - stats, various places where we don't have the txn's context anymore (asynchronous abort, resendWithTxn).

---

*[kv/txn_coord_sender.go, line 272 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8625#-KPTKCEpnPo2LxFy7yZP:-KPTNdmT5ErcS_JYqwcf:bsve0or) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca6615357cff9ed65c813b50d79c60d75b4023f8/kv/txn_coord_sender.go#L272)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> so you use the span from one context and the tracer from another. Is that cool?
</details>
No, but this is what the old code did and I am trying to make incremental changes. In practice it's probably all the same Tracer. The aim is to always use `ctx`, hence the TODO I left.

---

*[kv/txn_coord_sender.go, line 590 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8625#-KPTLLhrvpniwcQSZYpr:-KPTOMu1bU0uAKJRtDRr:b3bqh49) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca6615357cff9ed65c813b50d79c60d75b4023f8/kv/txn_coord_sender.go#L590)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> shouldn't this use the `ctx` method arg?
</details>
I don't think this function gets passed a useful context at this time (see the various TODOs above)

---

*[kv/txn_coord_sender.go, line 667 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8625#-KPTLDRQ2sDr7j7GPyhn:-KPTO6eQ6jZ98ydBFpYE:bj3h7vw) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca6615357cff9ed65c813b50d79c60d75b4023f8/kv/txn_coord_sender.go#L667)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> shouldn't `tryAsyncAbort` be passed in a context related to the txn it's working on?
</details>
I don't know the code well enough.. Maybe that txn's context isn't available anymore at this point?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8625#-:-KPTMmq-oMBUori33eoG:bfokfuo)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 0 of 11 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[kv/txn_coord_sender.go, line 174 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8625#-KPTJ_KjE3YzNrvDmV-l:-KPTPV2TF0-lo0N95BPe:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca6615357cff9ed65c813b50d79c60d75b4023f8/kv/txn_coord_sender.go#L174)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> check the ctx to not be cancellable too?
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8625)*
<!-- Sent from Reviewable.io -->

RaduBerinde committed:
kv: plumb context to TxnCoordSender

Part of #7993.

RaduBerinde committed:
storage: plumb context to node and store

Plumbing a context to Node and Store, and using it for logs and for the Tracer.

Part of #7994.

RaduBerinde committed:
kv: use a base context for DistSender Tracer and log messages

Pass a base context for DistSender. The Tracer should be set up in the context.

Part of #7993.
